src/txn/txn.c

```diff
@@ -895,8 +893,33 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
      * will be only one uncommitted prepared update. There can be a false positive of fixing history
      * store when handling prepared inserts, but it doesn't cost much.
      */
-    if (!F_ISSET(S2C(session), WT_CONN_IN_MEMORY) && resolved && upd == NULL)
-        WT_RET_NOTFOUND_OK(__txn_fixup_prepared_update(session, op, *cursorp, commit));
+    if (!F_ISSET(S2C(session), WT_CONN_IN_MEMORY) && resolved && upd == NULL) {
+        ret = __txn_fixup_prepared_update(session, op, *cursorp, commit);
+        /*
+         * If the above returns not found it means there wasn't a history store entry associated
+         * with the key, however it is still possible we wrote the prepared transaction to disk. In
+         * the event of a rollback we need to do some extra work.
+         */
+        if (ret == WT_NOTFOUND) {
+            if (!commit && F_ISSET(head_upd, WT_UPDATE_PREPARE_RESTORED_FROM_DISK)) {
+                cbt = (WT_CURSOR_BTREE *)*cursorp;
+
+                /*
+                 * Allocate a tombstone so that when we reconcile the update chain we don't copy the
+                 * prepared cell, which is now associated with a rolled back prepare, and instead
+                 * write nothing.
+                 */
+                WT_RET(__wt_upd_alloc_tombstone(session, &tombstone, &size));
+                /* Apply the tombstone to the row. */
+                WT_WITH_BTREE(session, op->btree, ret = __wt_row_modify(cbt, &cbt->iface.key, NULL,
```

Comment:
This is wrong, I think. We can only append at the end of the update chain not at head. See WT-6228

Reply:
I think once WT-6235 goes in I'll move it to be ahead of the point at which we abort the update, then I don't think we can race. If you think we still could, please explain how.

Reply:
May be that's safe, but I still want to append at the end. It feels natural to me as the updates do occur before the prepared updates and it is potentially safer. Imagine there is already a reader on the update chain, if we append at the end, we can reduce its chance to restart the read again.

Reply:
Conceptually it makes more sense to me that we put it ahead of the prepared update, in essence deleting everything that comes after it. I think the situation being discussed isn't likely to occur enough to justify appending to the end. In regards to updates occurring prior to the prepared update, we know that in this scenario that isn't possible.

Reply:
I think we can't agree on this. @bvpvamsikrishna @hbokhari What do you guys think?

Reply:
terms of before and after are confusing for me. We need to add tombstone at the head of the update chain, considering the newest will be at the head. This is the natural order.